### PR TITLE
Integrate with ACA's packet in; added throttling for sending packet ins.

### PR DESCRIPTION
**Pull request by zzxgzgz**
That this PR does:

1. Integrated with ACA's packet_in code path.
2. Added throttling to packet_in sending.

Building and Usage:
Please refer to [this link](https://floodlight.atlassian.net/wiki/spaces/floodlightcontroller/pages/1343657/Cbench+New) and [this link](https://opennetworkingfoundation.github.io/libfluid/md_doc_Benchmarks.html)

### CODE REVIEW

File: cbench/fakeswitch.c [hunk 1]
```diff
@@ -88,18 +88,22 @@ void fakeswitch_learn_dstmac(struct fakeswitch *fs)
...
-        0x29, 0x1a, 0x29, 0x1a, 0x08, 0x06, 0x00, 0x01, 
-        0x08, 0x00, 0x06, 0x04, 0x00, 0x02, 0x00, 0x0c, 
-        0x29, 0x1a, 0x29, 0x1a, 0x7f, 0x00, 0x00, 0x01, 
+        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x0c, // 8 octects per line
+                                // Ethernet Type: 802.1Q VLAN tagging (0x8100)
+        0x29, 0x1a, 0x29, 0x1a, 0x81, 0x00, 0x00, 0x01, //<- this 0x00, 0x01 (which is 1) is the vlan_id
+        // Ethernet Type: ARP (0x0806) 
+        0x08, 0x06, 0x06, 0x04, 0x00, 0x02, 0x00, 0x0c, 
+        // ARP OP Code: 1 is request, 2 is reply
+        0x00, 0x01, 0x29, 0x1a, 0x7f, 0x00, 0x00, 0x01,
```
Comment (xieus) [inline cbench/fakeswitch.c:97]: Is this part fixed or tunable?

File: cbench/fakeswitch.c [hunk 2]
```diff
@@ -376,20 +392,26 @@ void fakeswitch_handle_read(struct fakeswitch *fs)
             struct ofp_stats_request * stats_req;
             case OFPT_PACKET_OUT:
                 po = (struct ofp_packet_out *) ofph;
+                debug_msg(fs, "Got OFPT_PACKET_OUT, switch_status: %ld", fs->switch_status);
+                
                 if ( fs->switch_status == READY_TO_SEND && ! packet_out_is_lldp(po)) { 
-                    // assume this is in response to what we sent
+                    assume this is in response to what we sent
```
Comment (xieus) [inline cbench/fakeswitch.c]: Hey @zzxgzgz, will uncommenting line 398 fail the compile?

**Review by xieus — COMMENTED** [review]
@zzxgzgz Two comments:
(1) could you highlight the throttling part as this is the main payload;
(2) does Cbench have some CI workflow we could leverage so that we could avoid missing compile error?

File: cbench/fakeswitch.c [hunk 3]
```diff
@@ -376,20 +392,26 @@ void fakeswitch_handle_read(struct fakeswitch *fs)
...
+                }else{
+                    debug_msg(fs, "Got OFPT_PACKET_OUT but it doesn't count");
                 }
                 break;
             case OFPT_FLOW_MOD:
-                fm = (struct ofp_flow_mod *) ofph;
-                if(fs->switch_status == READY_TO_SEND && (fm->command == htons(OFPFC_ADD) || 
-                        fm->command == htons(OFPFC_MODIFY_STRICT)))
-                {
+                // fm = (struct ofp_flow_mod *) ofph;
```
Comment (lfu-ps) [inline cbench/fakeswitch.c]: I think we can uncomment/revert this part of code, and verify if the counter is correctly added next time for a Flow-Mod stress test

File: cbench/fakeswitch.c [hunk 4]
```diff
@@ -401,7 +423,7 @@ void fakeswitch_handle_read(struct fakeswitch *fs)
                 fakeswitch_change_status(fs, fs->learn_dstmac ? LEARN_DSTMAC : READY_TO_SEND);
                 break;
             case OFPT_SET_CONFIG:
-                // pull msgs out of buffer
+                // pull msgs out of buffer1
```
Comment (lfu-ps) [inline cbench/fakeswitch.c]: why 'buffer1' ?

File: cbench/fakeswitch.c (hunk 2 above)
Comment (lfu-ps) [reply cbench/fakeswitch.c]: agree, @zzxgzgz  please fix

Comment (zzxgzgz) [reply cbench/fakeswitch.c]: Fixed.

File: cbench/fakeswitch.c (hunk 3 above)
Comment (zzxgzgz) [reply cbench/fakeswitch.c]: Reverted.

File: cbench/fakeswitch.c (hunk 4 above)
Comment (zzxgzgz) [reply cbench/fakeswitch.c]: should be a typo, fixed.

File: cbench/fakeswitch.c (hunk 1 above)
Comment (zzxgzgz) [reply cbench/fakeswitch.c:97]: This part we can customize it according to our future needs.

At the moment, the src ip(`ip_address_to_learn`) and destination ip(`destination_ip_address`) can be tuned.

**Comment by zzxgzgz** [comment]
> @zzxgzgz Two comments: (1) could you highlight the throttling part as this is the main payload; (2) does Cbench have some CI workflow we could leverage so that we could avoid missing compile error?

1. I updated the code in cbench.c so the throtlling part can be found when you search for the word `throttling`.
2. They don't have any CICD that I'm aware of, that is an old repo and they stopped updating it about 6 years ago.

**Review by lfu-ps — APPROVED** [review]

**Pull request merged**